src/base/scene.cu

```diff
@@ -0,0 +1,309 @@
+    size_t positionsBufferOffset = positionsBufferView.byteOffset;
+    size_t positionsBufferLength = positionsBufferView.byteLength;
+    // TODO handle byteStride
+    Vector3f *positionsBytes = reinterpret_cast<Vector3f *>(
+        positionsBuffer.data.data() + positionsBufferOffset);
+    std::vector<Vector3f> primitivePositions(
+        positionsBytes,
+        positionsBytes + positionsBufferLength / sizeof(Vector3f));
+
+    // Extract normals
+    size_t normalsBufferOffset = normalsBufferView.byteOffset;
+    size_t normalsBufferLength = normalsBufferView.byteLength;
+    // TODO handle byteStride
+    Vector3f *normalsBytes = reinterpret_cast<Vector3f *>(
+        normalsBuffer.data.data() + normalsBufferOffset);
+    std::vector<Vector3f> primitiveNormals(
+        normalsBytes, normalsBytes + normalsBufferLength / sizeof(Vector3f));
+
+    // Extract face indices
+    size_t facesBufferOffset = facesBufferView.byteOffset;
+
+    static const size_t faceCount =
+        facesAccessor.count /
+        TRIANGLE_VERT_COUNT;  // gltf only supports triangles. No quads or
+                              // ngons
+
+    std::vector<Vector3i> primitiveFaces(faceCount);
+
+    // TODO handle byteStride
+
+    if (facesAccessor.componentType == 5123) {  // Component type unsigned short
+      unsigned short *facesBytes = reinterpret_cast<unsigned short *>(
+          facesBuffer.data.data() + facesBufferOffset);
+      for (size_t f = 0; f < faceCount; ++f) {
+        primitiveFaces[f] =
+            Vector3i(facesBytes[f * TRIANGLE_VERT_COUNT] + primitiveOffset,
+                     facesBytes[f * TRIANGLE_VERT_COUNT + 1] + primitiveOffset,
+                     facesBytes[f * TRIANGLE_VERT_COUNT + 2] + primitiveOffset);
+      }
+
+    } else if (facesAccessor.componentType ==
+               5125) {  // Component type unsigned int
+      unsigned int *facesBytes = reinterpret_cast<unsigned int *>(
+          facesBuffer.data.data() + facesBufferOffset);
+      for (size_t f = 0; f < faceCount; ++f) {
+        primitiveFaces[f] =
+            Vector3i(facesBytes[f * TRIANGLE_VERT_COUNT] + primitiveOffset,
+                     facesBytes[f * TRIANGLE_VERT_COUNT + 1] + primitiveOffset,
+                     facesBytes[f * TRIANGLE_VERT_COUNT + 2] + primitiveOffset);
+      }
+    }
+
+    positions.insert(positions.end(), primitivePositions.begin(),
+                     primitivePositions.end());
+    normals.insert(normals.end(), primitiveNormals.begin(),
+                   primitiveNormals.end());
+    faces.insert(faces.end(), primitiveFaces.begin(), primitiveFaces.end());
+
+    primitiveOffset += primitivePositions.size();
+  }
+
+  return std::shared_ptr<Mesh>(new Mesh(positions, normals, faces));
+}
+
+std::shared_ptr<gm::PerspectiveCamera> gm::Scene::loadCamera(
+    const std::vector<int> &node_ids, const tinygltf::Model &model) {
+  // NOTE The following section uses raw pointers to improve performance and
```

Comment:
This is a bit of a false statement as far as I know, since smart pointers are designed to be zero-cost abstractions. But this does not matter right, so just keep things as is